RUMM-1246: Update generators code to be aligned with changes in dd-sdk-reactnative repo

0xnm opened the pull request:
### What does this PR do?

We need to follow-up on the changes made here https://github.com/DataDog/dd-sdk-reactnative/pull/32 and  here https://github.com/DataDog/dd-sdk-reactnative/pull/35. This change updates generator in the following manner: bridge related classes are declared in `dd-foundation.tsx` file now, and this file exports them further.

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [x] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

0xnm commented on generators/gen_reactnative.py:
```diff
@@ -117,6 +119,29 @@ def generate(self, definitions: list):
 
     def _generate_ts_index(self, definitions: list):
         output_path = prepare_output_path(self.output_folder, OUTPUT_TS_FOLDER, OUTPUT_TS_INDEX)
+        predicate = lambda definition: (definition['name'] not in ["DdSdk", "DdSdkConfiguration"])
```
looks a bit ugly, but otherwise we need to add some flag to schema.

mariusc83 commented on tests/gen_reactnative/test_structure/expected/android/src/main/java/com/datadog/reactnative/DataStructureOptionalExt.kt:
```diff
@@ -22,7 +22,7 @@ fun ReadableMap.asDataStructureOptional(): DataStructureOptional {
 }
 
-fun DataStructureOptional.toReadableMap(): WritableNativeMap {
+fun DataStructureOptional.toReadableMap(): ReadableMap {
```
nice...I also noticed this at some point.

mariusc83 commented on generators/gen_reactnative.py:
```diff
@@ -117,6 +119,29 @@ def generate(self, definitions: list):
 
     def _generate_ts_index(self, definitions: list):
         output_path = prepare_output_path(self.output_folder, OUTPUT_TS_FOLDER, OUTPUT_TS_INDEX)
+        predicate = lambda definition: (definition['name'] not in ["DdSdk", "DdSdkConfiguration"])
```
I guess at least we could extract those hardcoded values into some static constants or maybe just create a static set : `NOT_EXPOSED_DEFINITIONS`

0xnm replied:
thanks! I also added a test to check if exclusion works correctly.

xgouchet commented on generators/gen_reactnative.py:
```diff
@@ -61,6 +62,8 @@
 
 INDEX_TSX_WARNING = "console.log(\"index.tsx was re-generated; make sure you include any missing code\")\n\n"
 
+NOT_EXPOSED_DEFINITIONS = ["DdSdk", "DdSdkConfiguration"]
```
I'm not in favor of having these hardcoded names here. Seems very hacky. Instead we could add a property `exposed`  in the schema.

0xnm replied:
it is a hack indeed, which solves a very particular problem. not sure how much it will be needed in future, but I can go for the schema field.

xgouchet commented on generators/gen_reactnative.py:
```diff
@@ -103,6 +106,7 @@ def __init__(self, output_folder: str):
 
     def generate(self, definitions: list):
         self._generate_ts_index(definitions)
+        self._generate_ts_dd_foundation(definitions)
```
What's the purpose of splitting things between index and dd_foundation? Also not a fan of the `dd_foundation` name which is not very generic (could be `foundation` even though it's not a very explicit name).

0xnm replied:
the split is needed because otherwise there is a `require` circular chain, which was fixed here https://github.com/DataDog/dd-sdk-reactnative/pull/35.

xgouchet commented on tests/gen_reactnative/test_combined/input.json:
```diff
@@ -24,6 +24,56 @@
   },
+  {
+    "type": "data",
+    "name": "DdSdkConfiguration",
```
Let's not use this name in our test, it would bring some confusion.

0xnm replied:
it was added to test the `export` exclusion in the generated files. but if we go for the `exposed` schema field, it is not really needed to have that name, yes.

xgouchet commented on tests/gen_reactnative/test_combined/input.json:
```diff
@@ -24,6 +24,56 @@
+  },
+  {
+    "type": "interface",
+    "name": "DdSdk",
```
Let's not use this name in our test, it would bring some confusion.

0xnm commented:
@xgouchet I implemented the following changes:

* Added `exposed` field to the schema to get rid of hardcoded names.
* Renamed `dd-foundation.tsx` to `foundation.tsx`. If we agree on the name I will do similar update in `dd-sdk-reactnative` repo. Such name (with `foundation`) was chosen because file holds all the base classes exposed via native bridge.

xgouchet reviewed (approved):
Really Nice work :)
